File: 2026-08-05-braid/static/rga.js
// rga.js — a hand-ported, 1:1 mirror of crdt/rga.py's RGADocument.
//
// This is NOT a thin client that trusts the server: every browser tab
// running this file is a full, independent CRDT replica. It applies its
// own edits instantly (no round trip), and merges remote ops using the
// exact same deterministic placement rule as the Python engine — parity
// between the two is checked by tests/test_parity.py, which feeds the
// same op stream through both and asserts identical output.
//
// See crdt/rga.py for the algorithm's rationale; comments here focus on
// JS-specific translation notes.

function idOf(x) {
  if (x === null || x === undefined) return null;
  return [x[0], x[1]];
}

function idKey(id) {
  // ids are [counter, siteId]; used as Map keys, which need a primitive.
  return id === null ? "root" : `${id[0]}:${id[1]}`;
}

function idGreater(a, b) {
  if (a[0] !== b[0]) return a[0] > b[0];
  return a[1] > b[1]; // string comparison, same as Python's tuple compare
}

class RGADocument {
  constructor(siteId) {
    this.siteId = siteId;
    this.counter = 0;
    this.sequence = [];      // ordered array of node objects, tombstones included
    this.byId = new Map();   // idKey -> node
    this.index = new Map();  // idKey -> current index in `sequence`
  }

  _nextId() {
    this.counter += 1;
    return [this.counter, this.siteId];
  }

  _reindexFrom(start) {
    for (let i = start; i < this.sequence.length; i++) {
      this.index.set(idKey(this.sequence[i].id), i);
    }
  }

  _integrate(node) {
    const origin = node.origin;
    const leftIdx = origin !== null ? this.index.get(idKey(origin)) : -1;
    let i = leftIdx + 1;
    const n = this.sequence.length;
    while (i < n) {
      const other = this.sequence[i];
      const otherOrigin = other.origin;
      const otherOriginIdx = otherOrigin !== null ? this.index.get(idKey(otherOrigin)) : -1;
      if (otherOriginIdx < leftIdx) break;
      if (otherOriginIdx === leftIdx) {
        if (idGreater(other.id, node.id)) {
          i += 1;
          continue;
        }
        break;
      }
      i += 1;
    }
    this.sequence.splice(i, 0, node);
    this.byId.set(idKey(node.id), node);
    this._reindexFrom(i);
  }

  _originForVisibleIndex(index) {
    if (index === 0) return null;
    let count = 0;
    for (const node of this.sequence) {
      if (!node.deleted) {
        count += 1;
        if (count === index) return node.id;
      }
    }
    throw new RangeError(`insert index ${index} out of range (doc length ${this.visibleLength()})`);
  }

  _visibleNodeAt(index) {
    let count = -1;
    for (const node of this.sequence) {
      if (!node.deleted) {
        count += 1;
        if (count === index) return node;
      }
    }
    throw new RangeError(`delete index ${index} out of range (doc length ${this.visibleLength()})`);
  }

  localInsert(index, value) {
    if (value.length !== 1) throw new Error("localInsert takes exactly one character");
    const origin = this._originForVisibleIndex(index);
    const nodeId = this._nextId();
    const node = { id: nodeId, origin, value, deleted: false };
    this._integrate(node);
    return { type: "insert", id: nodeId, origin: origin, value };
  }

  localDelete(index) {
    const node = this._visibleNodeAt(index);
    node.deleted = true;
    const opId = this._nextId();
    return { type: "delete", id: opId, target: node.id };
  }

  // delete/restore ops get their OWN fresh id (not the target node's id):
  // the same node can be deleted, undone (restored), and deleted again
  // over its lifetime, and each of those needs to be a distinguishable
  // op — see rga.py's local_delete docstring for the bug this avoids.
  deleteById(nodeId) {
    const node = this.byId.get(idKey(nodeId));
    node.deleted = true;
    const opId = this._nextId();
    return { type: "delete", id: opId, target: node.id };
  }

  restoreById(nodeId) {
    const node = this.byId.get(idKey(nodeId));
    node.deleted = false;
    const opId = this._nextId();
    return { type: "restore", id: opId, target: node.id };
  }

  applyRemoteOp(op) {
    if (op.type === "insert") {
      const nodeId = idOf(op.id);
      if (this.byId.has(idKey(nodeId))) return true; // duplicate delivery
      const origin = idOf(op.origin);
      if (origin !== null && !this.byId.has(idKey(origin))) return false; // dependency not met
      const node = { id: nodeId, origin, value: op.value, deleted: false };
      this._integrate(node);
      return true;
    } else if (op.type === "delete") {
      const target = idOf(op.target);
      const node = this.byId.get(idKey(target));
      if (!node) return false;
      node.deleted = true;
      return true;
    } else if (op.type === "restore") {
      const target = idOf(op.target);
      const node = this.byId.get(idKey(target));
      if (!node) return false;
      node.deleted = false;
      return true;
    }
    throw new Error(`unknown op type: ${op.type}`);
  }

  toString() {
    return this.sequence.filter((n) => !n.deleted).map((n) => n.value).join("");
  }

  visibleLength() {
    let c = 0;
    for (const n of this.sequence) if (!n.deleted) c += 1;
    return c;
  }

  // Number of visible characters strictly before the node with this id —
  // i.e. the position it occupies in the visible string if it's visible
  // itself, or the position it *would* occupy if it were. Used by the
  // browser UI to keep the local caret sane across remote edits (see
  // app.js) without needing a full Yjs-style relative-position system.
  visibleIndexOfId(id) {
    const key = idKey(id);
    let count = 0;
    for (const node of this.sequence) {
      if (idKey(node.id) === key) return count;
      if (!node.deleted) count += 1;
    }
    return -1;
  }
}

// ---- Site: causal buffering + op log + undo/redo (mirrors crdt/site.py) --

class Site {
  constructor(siteId) {
    this.siteId = siteId;
    this.doc = new RGADocument(siteId);
    this.pending = [];
    this.opLog = [];
    this.appliedKeys = new Set();
    this.undoStack = [];
    this.redoStack = [];
  }

  static opKey(op) {
    return `${op.type}:${idKey(op.id)}`;
  }

  _recordLocal(op) {
    this.appliedKeys.add(Site.opKey(op));
    this.opLog.push(op);
  }

  typeInsert(index, value) {
    const op = this.doc.localInsert(index, value);
    this._recordLocal(op);
    this.undoStack.push({ kind: "insert", nodeIds: [op.id] });
    this.redoStack = [];
    return op;
  }

  typeDelete(index) {
    const op = this.doc.localDelete(index);
    this._recordLocal(op);
    // op.id is the delete op's own fresh identity, not the node it acts
    // on — the undo stack needs the *target* node id.
    this.undoStack.push({ kind: "delete", nodeIds: [op.target] });
    this.redoStack = [];
    return op;
  }

  // Insert a multi-character string (typing fast, or pasting) as a single
  // grouped undo action — one undo removes the whole run.
  typeText(index, text) {
    const ops = [];
    const nodeIds = [];
    for (let i = 0; i < text.length; i++) {
      const op = this.doc.localInsert(index + i, text[i]);
      this._recordLocal(op);
      ops.push(op);
      nodeIds.push(op.id);
    }
    if (nodeIds.length > 0) {
      this.undoStack.push({ kind: "insert", nodeIds });
      this.redoStack = [];
    }
    return ops;
  }

  // Delete `count` characters starting at `index`, grouped as one undo
  // action. Each deletion targets the same visible index since deleting
  // shifts the rest of the text left.
  typeDeleteRange(index, count) {
    const ops = [];
    const nodeIds = [];
    for (let i = 0; i < count; i++) {
      const op = this.doc.localDelete(index);
      this._recordLocal(op);
      ops.push(op);
      nodeIds.push(op.target);
    }
    if (nodeIds.length > 0) {
      this.undoStack.push({ kind: "delete", nodeIds });
      this.redoStack = [];
    }
    return ops;
  }

  undo() {
    if (this.undoStack.length === 0) return null;
    const entry = this.undoStack.pop();
    const compensate = entry.kind === "insert" ? (id) => this.doc.deleteById(id) : (id) => this.doc.restoreById(id);
    const ops = entry.nodeIds.map(compensate);
    ops.forEach((op) => this._recordLocal(op));
    this.redoStack.push(entry);
    return ops;
  }

  redo() {
    if (this.redoStack.length === 0) return null;
    const entry = this.redoStack.pop();
    const compensate = entry.kind === "insert" ? (id) => this.doc.restoreById(id) : (id) => this.doc.deleteById(id);
    const ops = entry.nodeIds.map(compensate);
    ops.forEach((op) => this._recordLocal(op));
    this.undoStack.push(entry);
    return ops;
  }

  canUndo() { return this.undoStack.length > 0; }
  canRedo() { return this.redoStack.length > 0; }

  receive(op) {
    if (this._tryApply(op)) this._drainPending();
    else this.buffer(op);
  }

  _tryApply(op) {
    const key = Site.opKey(op);
    if (this.appliedKeys.has(key)) return true;
    if (!this.doc.applyRemoteOp(op)) return false;
    this.appliedKeys.add(key);
    this.opLog.push(op);
    return true;
  }

  _drainPending() {
    if (this.pending.length === 0) return;
    let progressed = true;
    while (progressed && this.pending.length > 0) {
      progressed = false;
      const stillPending = [];
      for (const op of this.pending) {
        if (this._tryApply(op)) progressed = true;
        else stillPending.push(op);
      }
      this.pending = stillPending;
    }
  }

  buffer(op) {
    if (this.appliedKeys.has(Site.opKey(op))) return;
    this.pending.push(op);
  }

  text() { return this.doc.toString(); }
  isSettled() { return this.pending.length === 0; }
}

if (typeof module !== "undefined" && module.exports) {
  module.exports = { RGADocument, Site, idGreater, idKey };
}
if (typeof window !== "undefined") {
  window.BraidRGA = { RGADocument, Site, idGreater, idKey };
}

File: 2026-08-05-braid/crdt/rga.py
"""
rga.py — a from-scratch Replicated Growable Array (RGA) CRDT.

RGA (Roh, Jeon, Kim & Lee, 2011, "Replicated abstract data types: Building
blocks for collaborative applications") represents a sequence (here: the
characters of a text document) as a list of tombstone-capable nodes. Every
node has a globally-unique id `(counter, site_id)` and an `origin` — the id
of the node it was inserted immediately to the right of at the moment it
was created (or `None` for "the very start of the document").

The key trick that makes two independently-edited replicas converge to the
*same* sequence without ever talking to each other during the edit is the
`_integrate` placement rule below: when two sites concurrently insert right
after the same origin, every replica breaks the tie the same deterministic
way (by comparing ids), so however the two inserts arrive — in whichever
order, with whatever delay, even delivered twice — every replica ends up
placing them in the identical relative order.

Deletes are tombstones (`deleted=True`), never physically removed, because
a later, concurrently-created insert might still reference a deleted node
as its origin — removing the node would break that reference.

This module is deliberately self-contained (no threads, no network, no
wall-clock) so it can be tested as pure data-in/data-out, and so it has an
exact 1:1 JS port in static/rga.js for the browser.
"""

from __future__ import annotations


def _id_of(x):
    """Normalize an id to a hashable tuple; accepts a 2-list (from JSON/the
    network) or a 2-tuple (from local Python code) or None."""
    if x is None:
        return None
    return (x[0], x[1])


def _id_greater(a, b):
    """Deterministic total order over ids: higher counter wins; the
    site_id (a string) is the tiebreak for same-counter ids from different
    sites, which can genuinely happen (two sites both make their Nth local
    edit). Every replica — Python or JS — must use this exact rule."""
    if a[0] != b[0]:
        return a[0] > b[0]
    return a[1] > b[1]


class RGADocument:
    """A single site's replica of a shared text document."""

    def __init__(self, site_id):
        self.site_id = site_id
        self._counter = 0
        self.sequence = []          # ordered list of node dicts, tombstones included
        self.by_id = {}             # id tuple -> node dict (same object as in `sequence`)
        self._index = {}            # id tuple -> current index in `sequence`

    # ---- id allocation -----------------------------------------------

    def _next_id(self):
        self._counter += 1
        return (self._counter, self.site_id)

    # ---- the RGA placement algorithm ----------------------------------

    def _reindex_from(self, start):
        for i in range(start, len(self.sequence)):
            self._index[self.sequence[i]["id"]] = i

    def _integrate(self, node):
        """Insert `node` (already carrying its final id/origin/value) into
        `self.sequence` at the position the RGA algorithm dictates, given
        whatever nodes are already present locally."""
        origin = node["origin"]
        left_idx = self._index[origin] if origin is not None else -1
        i = left_idx + 1
        n = len(self.sequence)
        while i < n:
            other = self.sequence[i]
            other_origin = other["origin"]
            other_origin_idx = self._index[other_origin] if other_origin is not None else -1
            if other_origin_idx < left_idx:
                # `other` descends from something left of our own origin —
                # we've walked past our whole "insert zone"; stop here.
                break
            if other_origin_idx == left_idx:
                # A direct sibling: another node also inserted immediately
                # after our origin. Break the tie deterministically —
                # the higher id stays closer to the origin.
                if _id_greater(other["id"], node["id"]):
                    i += 1
                    continue
                break
            # other_origin_idx > left_idx: `other` is nested under a
            # sibling (or a descendant of one) that sits to our right —
            # skip over that whole subtree, it doesn't compete with us.
            i += 1
        self.sequence.insert(i, node)
        self.by_id[node["id"]] = node
        self._reindex_from(i)

    # ---- local edits (this site typing) --------------------------------

    def _origin_for_visible_index(self, index):
        if index == 0:
            return None
        count = 0
        for node in self.sequence:
            if not node["deleted"]:
                count += 1
                if count == index:
                    return node["id"]
        raise IndexError(f"insert index {index} out of range (doc length {self.visible_length()})")

    def _visible_node_at(self, index):
        count = -1
        for node in self.sequence:
            if not node["deleted"]:
                count += 1
                if count == index:
                    return node
        raise IndexError(f"delete index {index} out of range (doc length {self.visible_length()})")

    def local_insert(self, index, value):
        """Insert a single character `value` at visible position `index`.
        Returns the op (a plain, JSON-serializable dict) to broadcast."""
        if len(value) != 1:
            raise ValueError("local_insert takes exactly one character")
        origin = self._origin_for_visible_index(index)
        node_id = self._next_id()
        node = {"id": node_id, "origin": origin, "value": value, "deleted": False}
        self._integrate(node)
        return {
            "type": "insert",
            "id": list(node_id),
            "origin": list(origin) if origin is not None else None,
            "value": value,
        }

    def local_delete(self, index):
        """Tombstone the character currently at visible position `index`.
        Returns the op to broadcast.

        Note that the returned op's own `id` is a *freshly allocated* id,
        distinct from the `target` node's id — a delete/restore op needs
        its own identity because the same node can legitimately be
        deleted, undone (restored), and deleted again over its lifetime
        (see undo/redo below); if a delete op reused its target's id, the
        second delete of the same node would be indistinguishable from a
        duplicate network delivery of the first one and get silently
        dropped by the dedup logic in site.py."""
        node = self._visible_node_at(index)
        node["deleted"] = True
        op_id = self._next_id()
        return {"type": "delete", "id": list(op_id), "target": list(node["id"])}

    # ---- id-targeted edits (used by undo/redo, not the live editor) ----
    #
    # A normal edit targets a *visible position* because that's what typing
    # means. Undo is different: "undo the character I inserted 47 edits
    # ago" has to keep meaning the same physical character no matter how
    # much the document has moved around since — including edits from
    # other peers landing in between. So undo/redo target a node **id**
    # directly rather than a position, and are safe no-ops if the node is
    # already in the state being asked for (e.g. undoing your insert after
    # someone else already deleted that same character).

    def delete_by_id(self, node_id):
        node = self.by_id[_id_of(node_id)]
        node["deleted"] = True
        op_id = self._next_id()
        return {"type": "delete", "id": list(op_id), "target": list(node["id"])}

    def restore_by_id(self, node_id):
        node = self.by_id[_id_of(node_id)]
        node["deleted"] = False
        op_id = self._next_id()
        return {"type": "restore", "id": list(op_id), "target": list(node["id"])}

    # ---- applying ops received from other sites ------------------------

    def apply_remote_op(self, op):
        """Apply an op produced by (a possibly different) `local_insert` /
        `local_delete` / this same method elsewhere. Returns True if the
        op was applied, False if it can't be applied yet because its
        causal dependency (the origin node for an insert, or the target
        node for a delete) hasn't arrived at this replica yet — the
        caller is expected to buffer and retry such ops. Applying the
        same insert op twice is a safe no-op (idempotent), which matters
        because the network may duplicate messages."""
        if op["type"] == "insert":
            node_id = _id_of(op["id"])
            if node_id in self.by_id:
                return True  # already have it — duplicate delivery
            origin = _id_of(op["origin"])
            if origin is not None and origin not in self.by_id:
                return False  # causal dependency not met yet
            node = {"id": node_id, "origin": origin, "value": op["value"], "deleted": False}
            self._integrate(node)
            return True
        elif op["type"] == "delete":
            target = _id_of(op["target"])
            node = self.by_id.get(target)
            if node is None:
                return False  # the insert that created the target hasn't arrived yet
            node["deleted"] = True
            return True
        elif op["type"] == "restore":
            # the undo/redo counterpart to "delete": un-tombstone a node.
            # Same causal-dependency rule as delete — buffer until the
            # insert that created the target node has arrived.
            target = _id_of(op["target"])
            node = self.by_id.get(target)
            if node is None:
                return False
            node["deleted"] = False
            return True
        raise ValueError(f"unknown op type: {op['type']!r}")

    # ---- reading ---------------------------------------------------------

    def to_string(self):
        return "".join(n["value"] for n in self.sequence if not n["deleted"])

    def visible_length(self):
        return sum(1 for n in self.sequence if not n["deleted"])

    def snapshot(self):
        """A deep, JSON-safe snapshot of every node (including tombstones)
        for debugging / the web UI's raw-state inspector."""
        return [
            {"id": list(n["id"]), "origin": list(n["origin"]) if n["origin"] else None,
             "value": n["value"], "deleted": n["deleted"]}
            for n in self.sequence
        ]
